File: src/main/java/lsd/cucumber/LsdCucumberPlugin.java
package lsd.cucumber;

import com.lsd.core.LsdContext;
import com.lsd.core.domain.Newpage;
import com.lsd.core.domain.PageTitle;
import io.cucumber.plugin.ConcurrentEventListener;
import io.cucumber.plugin.event.*;
import org.junit.platform.commons.util.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

import static io.cucumber.plugin.event.Status.FAILED;
import static java.util.stream.Collectors.joining;

public class LsdCucumberPlugin implements ConcurrentEventListener {

    private final LsdContext lsdContext = LsdContext.getInstance();

    private final List<TestCaseFinished> testCaseFinishedEvents = new ArrayList<>();
    private String scenarioName;
    private String featureName;

    private final Map<String, Integer> outlineScenarioNames = new ConcurrentHashMap<>();

    @Override
    public void setEventPublisher(EventPublisher publisher) {
        publisher.registerHandlerFor(TestCaseStarted.class, this::handleTestCaseStarted);
        publisher.registerHandlerFor(TestStepStarted.class, this::handleTestStepStarted);
        publisher.registerHandlerFor(TestCaseFinished.class, testCaseFinishedEvents::add);
        publisher.registerHandlerFor(TestRunFinished.class, this::handleTestRunFinished);
    }

    private void handleTestStepStarted(TestStepStarted testStepStarted) {
        var testStep = testStepStarted.getTestStep();
        if (testStep instanceof PickleStepTestStep) {
            var pickleTestStep = (PickleStepTestStep) testStep;
            var step = pickleTestStep.getStep();
            var stepText = step.getText();
            var stepKeyword = step.getKeyword();
            lsdContext.capture(new Newpage(new PageTitle(stepKeyword + stepText)));
        }
    }

    private void handleTestCaseStarted(TestCaseStarted event) {
        TestCase testCase = event.getTestCase();
        String currentFeatureName = retrieveFeatureName(testCase);
        String currentScenarioName = getScenarioName(testCase);
        if (isVeryFirstRun()) {
            prepareFirstScenario(currentFeatureName, currentScenarioName);
        } else if (!continuationOfExistingScenario(currentScenarioName)) {
            finishProcessingCompletedScenario();
            finishProcessingCompletedFeature(currentFeatureName);
            prepareForNewScenario(currentScenarioName, currentFeatureName);
        } else {
            finishProcessingCompletedScenario();
            prepareForNewScenario(currentScenarioName, currentFeatureName);
        }
    }

    private String getScenarioName(TestCase testCase) {
        if (testCase.getKeyword().equalsIgnoreCase("Scenario Outline")) {
            Integer index = Optional.ofNullable(outlineScenarioNames.get(testCase.getName())).orElse(0);
            outlineScenarioNames.put(testCase.getName(), ++index);
            return testCase.getName() + " #" + index;
        }
        return testCase.getName();
    }

    private String retrieveFeatureName(TestCase testCase) {
        return testCase.getUri().toString().replaceAll(".*/(.*?).feature$", "$1");
    }

    private void prepareFirstScenario(String currentFeatureName, String currentScenarioName) {
        scenarioName = currentScenarioName;
        featureName = currentFeatureName;
    }

    private void prepareForNewScenario(String currentScenarioName, String currentFeatureName) {
        testCaseFinishedEvents.clear();
        scenarioName = currentScenarioName;
        featureName = currentFeatureName;
    }

    private void finishProcessingCompletedScenario() {
        var description = testCaseFinishedEvents.stream()
                .flatMap(event -> event.getTestCase().getTestSteps().stream())
                .filter(PickleStepTestStep.class::isInstance)
                .map(PickleStepTestStep.class::cast)
                .map(v -> v.getStep().getKeyword() + " " + v.getStep().getText())
                .collect(joining("</br>"));


        var result = testCaseFinishedEvents.stream()
                .map(TestCaseFinished::getResult)
                .filter(testResult -> testResult.getStatus().is(FAILED))
                .findFirst()
                .map(x -> com.lsd.core.domain.Status.ERROR)
                .orElse(com.lsd.core.domain.Status.SUCCESS);

        lsdContext.completeScenario(scenarioName, description, result);
    }

    private void finishProcessingCompletedFeature(String currentFeatureName) {
        if (featureName != null && !currentFeatureName.equalsIgnoreCase(featureName)) {
            lsdContext.completeReport(featureName);
        }
    }

    private boolean continuationOfExistingScenario(String testScenarioName) {
        return testScenarioName.equals(scenarioName);
    }

    private boolean isVeryFirstRun() {
        return null == scenarioName;
    }

    private void handleTestRunFinished(TestRunFinished event) {
        if (!StringUtils.isBlank(featureName)) {
            finishProcessingCompletedScenario();
            lsdContext.completeReport(featureName);
            lsdContext.createIndex();
            lsdContext.completeComponentsReport("Combined Component Diagram");
        }
    }
}

File: src/main/kotlin/cucumber/LsdCucumberPlugin.kt
package lsd.cucumber

import com.lsd.core.LsdContext
import com.lsd.core.ReportOptions
import com.lsd.core.domain.Newpage
import com.lsd.core.domain.PageTitle
import com.lsd.core.domain.Status.ERROR
import com.lsd.core.domain.Status.SUCCESS
import com.lsd.core.properties.LsdProperties.getBoolean
import io.cucumber.plugin.ConcurrentEventListener
import io.cucumber.plugin.event.*
import io.cucumber.plugin.event.Status.FAILED
import java.util.concurrent.ConcurrentHashMap

class LsdCucumberPlugin : ConcurrentEventListener {
    private val lsdContext: LsdContext = LsdContext.instance
    private val testCaseFinishedEvents: MutableList<TestCaseFinished> = ArrayList()
    private val options = ReportOptions()
    private val outlineScenarioNames: MutableMap<String, Int> = ConcurrentHashMap()

    private var scenarioName: String? = null
    private var featureName: String? = null
    private var splitByStep: Boolean = false

    override fun setEventPublisher(publisher: EventPublisher) {
        splitByStep = getBoolean("lsd.cucumber.splitBySteps")
        publisher.registerHandlerFor(TestCaseStarted::class.java, ::handleTestCaseStarted)
        publisher.registerHandlerFor(TestStepStarted::class.java, ::handleTestStepStarted)
        publisher.registerHandlerFor(TestCaseFinished::class.java, testCaseFinishedEvents::add)
        publisher.registerHandlerFor(TestRunFinished::class.java, ::handleTestRunFinished)
    }

    private fun handleTestStepStarted(testStepStarted: TestStepStarted) {
        val testStep = testStepStarted.testStep
        if (splitByStep && testStep is PickleStepTestStep) {
            val step = testStep.step
            lsdContext.capture(Newpage(PageTitle(step.keyword + step.text)))
        }
    }

    private fun handleTestCaseStarted(event: TestCaseStarted) {
        val testCase = event.testCase
        val currentFeatureName = testCase.featureName()
        val currentScenarioName = testCase.scenarioName()
        if (isVeryFirstRun) {
            prepareFirstScenario(currentFeatureName, currentScenarioName)
        } else if (!continuationOfExistingScenario(currentScenarioName)) {
            finishProcessingCompletedScenario()
            finishProcessingCompletedFeature(currentFeatureName)
            prepareForNewScenario(currentScenarioName, currentFeatureName)
        } else {
            finishProcessingCompletedScenario()
            prepareForNewScenario(currentScenarioName, currentFeatureName)
        }
    }

    private fun TestCase.scenarioName(): String {
        if (keyword.equals("Scenario Outline", ignoreCase = true)) {
            var index = outlineScenarioNames[name] ?: 0
            outlineScenarioNames[name] = ++index
            return "$name #$index"
        }
        return name
    }

    private fun TestCase.featureName(): String =
        uri.toString().replace(".*/(.*?).feature$".toRegex(), "$1")

    private fun prepareFirstScenario(currentFeatureName: String, currentScenarioName: String) {
        scenarioName = currentScenarioName
        featureName = currentFeatureName
    }

    private fun prepareForNewScenario(currentScenarioName: String, currentFeatureName: String) {
        testCaseFinishedEvents.clear()
        scenarioName = currentScenarioName
        featureName = currentFeatureName
    }

    private fun finishProcessingCompletedScenario() {
        val description = testCaseFinishedEvents
            .flatMap { it.testCase.testSteps }
            .filterIsInstance<PickleStepTestStep>()
            .joinToString(separator = "</br>") { "${it.step.keyword} ${it.step.text}" }

        val overallStatus =
            if (testCaseFinishedEvents.any { it.result.status == FAILED }) ERROR else SUCCESS

        lsdContext.completeScenario(scenarioName!!, description, overallStatus)
    }

    private fun finishProcessingCompletedFeature(currentFeatureName: String) {
        if (featureName != null && !currentFeatureName.equals(featureName, ignoreCase = true)) {
            lsdContext.completeReport(featureName!!)
        }
    }

    private fun continuationOfExistingScenario(testScenarioName: String): Boolean =
        testScenarioName == scenarioName

    private val isVeryFirstRun: Boolean
        get() = scenarioName == null

    private fun handleTestRunFinished(event: TestRunFinished) {
        featureName?.let { name ->
            if (name.isNotBlank()) {
                finishProcessingCompletedScenario()
                lsdContext.completeReport(name)
                lsdContext.createIndex(options.devMode)
                lsdContext.completeComponentsReport("Combined Component Diagram")
            }
        }
    }
}
